arch/arc/core/fast_irq.S

```diff
@@ -101,9 +101,11 @@ firq_nest:
 	 * save original value of _ARC_V2_USER_SP and ilink into
 	 * the stack of interrupted context first, then restore them later
 	 */
-	st ilink, [sp]
-	lr ilink, [_ARC_V2_USER_SP]
+	/* push ilink */
 	st ilink, [sp, -4]
+	lr ilink, [_ARC_V2_USER_SP]
+	/* push _ARC_V2_USER_SP  */
+	st ilink, [sp, -8]
 	/* sp here is the sp of interrupted context */
 	sr sp, [_ARC_V2_USER_SP]
```

Comment:
While we're at it one other question: don't we want to modify `SP` value as well? I mean do exactly what `PUSH` would do as you mention in the commit message: `sp=sp-4; *sp= b;`?

Indeed that's only required if we may switch to another process right in this IRQ handler and later return to interrupted context. And in any case that will add clarity and simplify understanding... especially if we use macros for saving/restoring of AUX regs similarly to Linux' `PUSHAX()`/`POPAX()`, see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arc/include/asm/entry.h#n37.